### PR TITLE
Implement an API for performance monitoring

### PR DESCRIPTION
**Pull request by adodon2go**
See detailed future description in #213 

Design an implement an API which provides information about the performance of the zot server which could be used for debugging.

Let's look at:

    storage - how much space is used per repo (folder)
    storage - how much total space is used
    latency counters - min/max/average - per repo (folder) per specific time interval
    how many times an image was downloaded
    how many times has an image been uploaded to a repo (folder)

### CODE REVIEW

File: examples/config-metrics.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    },
+    "http": {
+        "address": "127.0.0.1",
+        "port": "8080"
+    },
+    "log": {
+        "level": "debug"
+    },
+    "extensions": {
+        "metrics": {
```
Comment (shimish2) [inline examples/config-metrics.json:14]: If we are including metrics feature in minimal binary than we should use metrics section on top level

Comment (rchincha) [reply examples/config-metrics.json:14]: I think there are basic in-memory metrics which are available.
"metrics" under extensions are the full prometheus implementation.
This is fine actually.

File: README.md [hunk 2]
```diff
@@ -41,6 +41,8 @@ https://anuvu.github.io/zot/
 * Swagger based documentation
 * Single binary for _all_ the above features
 * Released under Apache 2.0 License
+* Metric [instrumentation](#instrumentation) with Prometheus
```
Comment (rchincha) [inline README.md:44]: "Metrics"

File: README.md [hunk 3]
```diff
@@ -267,6 +269,21 @@ runtime interface.
 
 Works with "docker://" transport which is the default.
 
+# Instrumentation
```
Comment (rchincha) [inline README.md:272]: "Metrics"

Comment (rchincha) [reply README.md:272]: I will clean this up after your PR is merged.

File: pkg/exporter/api/config.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type ExporterConfig struct {
+	Port    string
+	Log     *LogConfig
+	Metrics *MetricsConfig
+}
+
+type Config struct {
+	ZotServer   ServerConfig
```
Comment (rchincha) [inline pkg/exporter/api/config.go:28]: Don't use project names for vars ("Zot...") 
What if we change the name of the project!

Comment (rchincha) [reply pkg/exporter/api/config.go:28]: Other places too!

File: pkg/api/routes.go [hunk 5]
```diff
@@ -96,8 +96,13 @@ func (rh *RouteHandler) SetupRoutes() {
 	// swagger swagger "/swagger/v2/index.html"
 	rh.c.Router.PathPrefix("/swagger/v2/").Methods("GET").Handler(httpSwagger.WrapHandler)
 	// Setup Extensions Routes
-	if rh.c.Config != nil && rh.c.Config.Extensions != nil {
-		ext.SetupRoutes(rh.c.Config, rh.c.Router, rh.c.StoreController, rh.c.Log)
+	if rh.c.Config != nil {
+		if rh.c.Config.Extensions == nil {
+			// minimal install
```
Comment (rchincha) [inline pkg/api/routes.go:101]: "minimal build"

File: pkg/api/routes.go [hunk 6]
```diff
@@ -96,8 +96,13 @@ func (rh *RouteHandler) SetupRoutes() {
...
+			g.HandleFunc("/metrics", rh.GetMetrics).Methods("GET")
+		} else {
+			ext.SetupRoutes(rh.c.Config, rh.c.Router, rh.c.StoreController, rh.c.Log)
```
Comment (rchincha) [inline pkg/api/routes.go:104]: "extended build"

File: pkg/exporter/api/config.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	ZotExporter ExporterConfig
+}
+
+func DefaultConfig() *Config {
+	return &Config{
+		ZotServer:   ServerConfig{Protocol: "http", Host: "localhost", Port: "8080"},
```
Comment (rchincha) [inline pkg/exporter/api/config.go:34]: What is this config for? If it is zot side, shouldn't you just read the zot server config?

File: pkg/extensions/monitoring/minimal_client.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	// Transport to use for the http client.
+	Transport *http.Transport
+
+	// HTTPClient is the client to use.
+	HTTPClient *http.Client
+}
+
+type MetricsClient struct {
+	headers http.Header
+	config  ZotMetricsConfig
```
Comment (rchincha) [inline pkg/extensions/monitoring/minimal_client.go:33]: MetricsConfig? instead

File: pkg/exporter/api/config.go (hunk 7 above)
Comment (adodon2go) [reply pkg/exporter/api/config.go:34]: This is not related to zot binary in any way. We have a zot-exporter binary with separate CLI options & config.

Comment (rchincha) [reply pkg/exporter/api/config.go]: Use more generic names such as RegistryServer or Exporter?

File: pkg/exporter/api/zot_exporter.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
+// +build minimal
```
Comment (rchincha) [inline pkg/exporter/api/zot_exporter.go:1]: Change this filename from zot_exporter.go to simply exporter.go

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**